# 4. Mixed Models

This tutorial extends what was previously introduced in the Single Variant Association Tests tutorial from regression models to mixed models that account for genetic relatedness among samples using the [GENESIS](https://bioconductor.org/packages/release/bioc/html/GENESIS.html) R/Bioconductor package. We use the output from the Population Structure and Relatedness Inference tutorial as inputs to our mixed models.

## Null Model

Recall that the first step in our association testing procedure is to fit the null model -- i.e. a model fit under the null hypothesis of no individual variant association. The difference from before is that we will fit a *mixed* model with the desired outcome phenotype, fixed effect covariates, *and a random effect with covariance proportional to the kinship matrix*.

#### Prepare the Data

In addition to the phenotype data that we used previously, we will also use the ancestry PCs and pairwise kinship estimates we created in the previous session. Recall that PCs 1-5 appear to reflect population structure in our sample, so we will use those to adjust for ancestry in our null model (i.e. the same PCs we used in our PC-Relate analysis). We need to add these PCs to our `AnnotatedDataFrame` with the phenotype data.

```{r null_model_mm, message = FALSE}
# sample annotation
repo_path <- "https://github.com/UW-GAC/SISG_2022/raw/main"
if (!dir.exists("data")) dir.create("data")
sampfile <- "data/sample_phenotype_annotation.RData"
if (!file.exists(sampfile)) download.file(file.path(repo_path, sampfile), sampfile)
annot <- get(load(sampfile))
library(Biobase)
head(pData(annot))

# load the ancestry PCs
pcfile <- "data/pcs.RData"
if (!file.exists(pcfile)) download.file(file.path(repo_path, pcfile), pcfile)
pcs <- get(load(pcfile))
pcs <- pcs[,c("sample.id", "PC1", "PC2", "PC3", "PC4", "PC5")]
head(pcs)

# merge PCs with the sample annotation
dat <- merge(pData(annot), pcs, by = "sample.id")
head(dat)

# update the AnnotatedDataFrame
pData(annot) <- dat
save(annot, file="data/sample_phenotype_pcs.RData")
```

We can create an (n x n) kinship matrix (KM) from the output of `pcrelate` using the `pcrelateToMatrix` function. We set `scaleKin = 2` to multiply the kinship values by 2, which gives values on the same scale as the standard GRM (this is relevant for the interpretation of the variance component estimates). This matrix is represented in R as a symmetric matrix object from the Matrix package.

```{r load_kinship}
kinfile <- "data/pcrelate.RData"
if (!file.exists(kinfile)) download.file(file.path(repo_path, kinfile), kinfile)
pcrel <- get(load(kinfile))

library(GENESIS)
kinship <- pcrelateToMatrix(pcrel, scaleKin=2, verbose=FALSE)
dim(kinship)
kinship[1:5,1:5]

# save output
save(kinship, file="data/pcrelate_Matrix.RData")
```

#### Fit the Null Model

When running a mixed model analysis, we still fit the null model using the `fitNullModel` function in GENESIS. Now, we include the kinship matrix in the model with the `cov.mat` (covariance matrix) argument, which is used to specify the random effect(s) in the model with covariance structure(s) proportional to the supplied matrix(s). The inclusion of these random effects is what makes this a mixed model, rather than a simple regression model. We also add the ancestry PCs to the list of covariates and allow for heterogeneous residual variance by `study` with the `group.var` argument, as before.

```{r null_model_fit_mm}
nullmod <- fitNullModel(annot, 
                        outcome="height", 
                        covars=c("sex", "age", "study", paste0("PC", c(1:5))), 
                        cov.mat=kinship, 
                        group.var="study", 
                        verbose=FALSE)

# save the output
save(nullmod, file="data/null_mixed_model_height.RData")
```

We can investigate the output from `fitNullModel`.
```{r}
# description of the model we fit
nullmod$model

# fixed effect regression estimates
nullmod$fixef

# variance component estimates by group.var
nullmod$varComp

# model fit: fitted values, residuals
head(nullmod$fit)

library(ggplot2)
ggplot(nullmod$fit, aes(x = fitted.values, y = resid.marginal)) + 
    geom_point(alpha = 0.5) + 
    geom_hline(yintercept = 0) + 
    geom_smooth(method = 'lm')
```

## Single-Variant Association Tests

Now we can run single-variant score tests to test for variant association with the outcome, accounting for genetic ancestry and genetic relatedness among the subjects. We use the same `assocTestSingle` function as before; the only difference is that we pass in our new null model. 

```{r assoc_single_mm, message = FALSE}
library(SeqVarTools)
gdsfile <- "data/1KG_phase3_subset_chr1.gds"
if (!file.exists(gdsfile)) download.file(file.path(repo_path, gdsfile), gdsfile)
gdsfmt::showfile.gds(closeall=TRUE) # make sure file is not already open
gds <- seqOpen(gdsfile)

# make the seqVarData object
seqData <- SeqVarData(gds, sampleData=annot)

# make the iterator object
iterator <- SeqVarBlockIterator(seqData, verbose=FALSE)

# run the single-variant association test
assoc <- assocTestSingle(iterator, 
                         null.model = nullmod)
dim(assoc)
head(assoc)
```

We make the usual QQ plot, filtering to variants with minor allele count (MAC) $\geq 5$.

```{r assoc_single_qq_mm}
library(ggplot2)
qqPlot <- function(pval) {
    pval <- pval[!is.na(pval)]
    n <- length(pval)
    x <- 1:n
    dat <- data.frame(obs=sort(pval),
                      exp=x/n,
                      upper=qbeta(0.025, x, rev(x)),
                      lower=qbeta(0.975, x, rev(x)))
    
    ggplot(dat, aes(-log10(exp), -log10(obs))) +
        geom_line(aes(-log10(exp), -log10(upper)), color="gray") +
        geom_line(aes(-log10(exp), -log10(lower)), color="gray") +
        geom_point() +
        geom_abline(intercept=0, slope=1, color="red") +
        xlab(expression(paste(-log[10], "(expected P)"))) +
        ylab(expression(paste(-log[10], "(observed P)"))) +
        theme_bw()
}    

qqPlot(assoc$Score.pval[assoc$MAC >= 5])
```

Notice that we observe much less inflation than before, when we did not adjust for ancestry and relatedness.


### Exercise 4.1 (Data Studio)

Perform a single-variant association test for `status`. Adjust for sex, age, study, ancestry, and kinship in the model. Don't forget to consider the `family` parameter. Make a QQ plot of the p-values for all variants with MAC >= 5.

```{r}
# your solution here 
#
#
#
#
#
#
#
#
#
```

### Solution 4.1 (Data Studio)

Perform a single-variant association test for `status`. Adjust for sex, age, study, ancestry, and kinship in the model. Don't forget to consider the `family` parameter. Make a QQ plot of the p-values for all variants with MAC >= 5.

The first step is to fit the null model. We include PCs 1-5 as covariates in our model to adjust for ancestry, and we include a random effect proportional to the kinship matrix to adjust for genetic relatedness. Recall that with a binary outcome, we set `family = binomial` -- because we have a random effect, this will fit an approximate logistic mixed model using the [GMMAT method](https://www.cell.com/ajhg/fulltext/S0002-9297(16)00063-X).

```{r exercise_mm_nullmod}
nullmod.status <- fitNullModel(annot, 
                               outcome="status", 
                               covars=c("sex", "age", "study", paste0("PC", c(1:5))), 
                               cov.mat=kinship, 
                               family = binomial,
                               verbose=FALSE)

# description of the model we fit
nullmod.status$model

# fixed effect regression estimates
nullmod.status$fixef

# variance component estimate for kinship
nullmod.status$varComp

# save the output
save(nullmod.status, file="data/null_mixed_model_status.RData")
```

Now that we have the null model, we perform the single-variant association tests and make the QQ plot the same way as before. 

```{r exercise_mm_assoc}
resetIterator(iterator, verbose=FALSE)
# run the single-variant association test
assoc.status <- assocTestSingle(iterator, 
                                null.model = nullmod.status, 
                                test = "Score")
dim(assoc.status)
head(assoc.status)

qqPlot(assoc.status$Score.pval[assoc.status$MAC >= 5])
```

```{r assoc_mm_close}
seqClose(seqData)
```


### Exercise 4.2 (Application)

Use the `GENESIS Null Model` app on the BioData Catalyst powered by Seven Bridges platform to fit the a null model for height, adjusting for sex, age, study, ancestry, and kinship in the model, using the example 1000 Genomes data. We use the PCs and kinship matrix we computed using the PC-AiR and PC-Relate apps in the Population Structure and Relatedness tutorial's exercises as inputs to this analysis. The steps to perform this analysis are as follows:

- Copy the app to your project if it is not already there:
  - Click: Public Gallery > Apps > Browse
  - Search for `GENESIS Null Model`
  - Click: Copy > Select your project > Copy
- Run the analysis in your project:
  - Click: Apps > `GENESIS Null Model` > Run
  - Specify the Inputs:
    - Phenotype file: `sample_phenotype_pcs.RData` 
    - Relatedness matrix file: `1KG_phase3_subset_pcrelate_Matrix.RData`
  - Specify the App Settings:
    - Covariates: sex, age, study, PC1, PC2, PC3, PC4, PC5, PC6 (each as a different element)
    - Family: gaussian
    - Group Variate: study
    - Outcome: height
    - Two stage model: FALSE
    - Output prefix: "1KG_height" (or any other string to name the output file)
  - Click: Run
  
The analysis will take a few minutes to run. You can find your analysis in the Tasks menu of your Project to check on its progress and see the results once it has completed.

The output of this analysis will be a `<output_prefix>_null_model.RData` file that contains the null model fit, a `<output_prefix>_phenotypes.RData` file with the phenotype data used in the analysis, and a `<output_prefix>_report.Rmd` and `<output_prefix>_report.html` with model diagnostics. Review the .html report -- which covariates have significant ($p < 0.05$) associations with height in the null model?

You can find the expected output of this analysis by looking at the existing task `06. 1KG Null Model` in the Tasks menu of your Project. The output files are available in the Project, so you do not need to wait for your analysis to finish to move to the next exercise.

### Solution 4.2 (Application)

From looking at the .html report, we see that sex, study, and PC2 have significant associations with height in our null model.


### Exercise 4.3 (Application)

Use the `GENESIS Single Variant Association Testing` app on the BioData Catalyst powered by Seven Bridges platform to perform a GWAS for height using the null model fit in the previous exercise. Use the genotype data in the genome-wide GDS files you created previously. The steps to perform this analysis are as follows:

- Copy the app to your project if it is not already there:
  - Click: Public Gallery > Apps > Browse
  - Search for `GENESIS Single Variant Association Testing`
  - Click: Copy > Select your project > Copy
- Run the analysis in your project:
  - Click: Apps > `GENESIS Single Variant Association Testing` > Run
  - Specify the Inputs:
    - GDS Files: `1KG_phase3_subset_chr<CHR>.gds` (select all 22 chromosomes)
    - Null model file: `1KG_height_null_model.RData`
    - Phenotype file: `1KG_height_phenotypes.RData` (use the phenotype file created by the Null Model app)
  - Specify the App Settings:
    - Genome build: hg19
    - MAC threshold: 5
    - memory GB: 32 (increase to make sure enough available)
    - Output prefix: "1KG_height_single" (or any other string to name the output file)
  - Click: Run
  
The analysis will take a few minutes to run. You can find your analysis in the Tasks menu of your Project to check on its progress and see the results once it has completed.

The output of this analysis will be 22 `<output_prefix>_single_chr<CHR>.RData` files with the association test results for each chromosome as well as a `<output_prefix>_single_manh.png` file with the Manhattan plot and a `<output_prefix>_single_qq.png` file with the QQ plot. Review the QQ and Manhattan plots -- is there evidence of genomic inflation?

You can find the expected output of this analysis by looking at the existing task `07. 1KG Single Variant Association Test` in the Tasks menu of your Project. The output files are available in the Project, so you do not need to wait for your analysis to finish to move to the next exercise.

### Solution 4.3 (Application)

From looking at the QQ plot, we see that the genomic control lambda = 1.127 and there is deviation from the $y=x$ line -- both indicative of inflation in our analysis. This is likely an artifact of looking at rare variants with a small sample size. 


### Exercise 4.4 (Application)

Use the `GENESIS Association results plotting` app on the BioData Catalyst powered by Seven Bridges platform to make additional QQ plots of the single variant association results binned by MAF: $0-0.5\%$, $0.5-1\%$, $1-5\%$, $\geq 5\%$. The steps to perform this analysis are as follows:

- Copy the app to your project if it is not already there:
  - Click: Public Gallery > Apps > Browse
  - Search for `GENESIS Association results plotting`
  - Click: Copy > Select your project > Copy
- Run the analysis in your project:
  - Click: Apps > `GENESIS Association results plotting` > Run
  - Specify the Inputs:
    - Results from association testing: `1KG_height_single_chr<CHR>.RData` (select all 22 chromosomes)
  - Specify the App Settings:
    - Association Type: single
    - QQ MAF bins: "0.005 0.01 0.05"
  - Click: Run

The analysis will take a few minutes to run. You can find your analysis in the Tasks menu of your Project to check on its progress and see the results once it has completed.

The output of this analysis will be a `<output_prefix>_single_qq_bymaf.png` file. Look at the QQ plots by MAF bin -- how do they compare to the overall QQ plot of all variants?

You can find the expected output of this analysis by looking at the existing task `08. 1KG Association Single Plots` in the Tasks menu of your Project, so you do not need to wait for your analysis to finish to look at the output.

### Solution 4.4 (Application)

From the binned QQ plots, we see that the common variants (i.e. MAF $\geq 5\%$) have a genomic control lambda = 1.047 and do not have significant deviation from the $y=x$ line. As suspected, the inflation is only present in the rarer variants, likely due to the small sample size. 
