# 5. Exploring Association Results

In this tutorial, we will learn how to use the [GENESIS Model Explorer App](https://genesis-model-explorer-app.bdc.sb-webapp.com/?project=smgogarten/uw-gac-commit) and the [LocusZoom Shiny App](https://locuszoom-shiny-app.bdc.sb-webapp.com/), which are both "Interactive Browsers" built with [R Shiny](https://shiny.rstudio.com/) on the NHLBI BioData Catalyst powered by Seven Bridges cloud platform. Currently, both of these applications cap the number of users at a given time, so we will not all be able to use it live concurrently during the workshop. The instructors will demonstrate the use of each tool by going through these exercises now, and everyone can perform the hands-on exercises after the workshop is over, at their own convenience.

## GENESIS Model Explorer App

The [GENESIS Model Explorer App](https://genesis-model-explorer-app.bdc.sb-webapp.com/?project=smgogarten/uw-gac-commit) is an interactive tool that enables users to make figures to visualize and explore the results of a GENESIS null model, paired with phenotype and genotype data on the same samples. It is meant to provide an intuitive interface for researchers to easily select, visualize, and explore phenotypes, genotypes, and a fitted GENESIS model interactively with no prior R programming knowledge. The app takes three inputs:

- **Null Model File:** The null model file should be any fitted GENESIS null model saved in .RData format. The null model could have been created interactively using the `fitNullModel` function in an R session (e.g. in Data Studio or on your local machine), or it could be the output from the `GENESIS Null Model` application.
- **Phenotype File:** The phenotype file should be a data.frame or `AnnotatedDataFrame` saved in .RData format. The data.frame must contain all of the samples included in your null model file in a column named `sample.id`, with additional columns containing phenotype variables of interest. If you used the `GENESIS Null Model` application to fit your null model, we recommend using the `<output_prefix>_phenotypes.RData` output file, which contains all of the phenotype data from all of the samples used in the analysis. Alternatively, you can use the same phenotype file used as input to fit your null model, or an entirely new file where you have added additional columns with phenotype variables of interest.
- **Genotype File (Optional):** Providing an optional genotype file allows the user to make figures looking at the relationships of variants of interest with null model variables and phenotypes of interest. The genotype file should be a data.frame saved in .rds format. The data.frame must contain all of the samples included in your null model file in a column named `sample.id`, with additional columns containing variant allele counts or dosages. Conveniently, this file can be generated from an existing GDS file with the `GDS Genotype Extractor` application (see below).

We will now use the [GENESIS Model Explorer](https://genesis-model-explorer-app.bdc.sb-webapp.com/?project=smgogarten/uw-gac-commit) to make some figures exploring the data:

- Launch the interactive browser
  - From the top menu, click "Public Resources" > "Interactive Web Apps"
  - Click: "Open" on the GENESIS Model Explorer App
  - Click: "Yes" to proceed
  - Click: "Get Started"
- Load Data
  - Null Model File
    - Project: select your SISG project (should be chosen by default if you launched the app as described)
    - Current File: select `1KG_height_null_model_reportonly.RData` (much smaller file without extra matrices required for computing association test statistics)
  - Phenotype File:
    - Project: select your SISG project
    - Current File: select `1KG_height_phenotypes.RData` (this is the phenotype file that was created by the null model application)
  - Click: Load Data

Once you load the data, you will be taken to a "Plot setup" screen, where you can select what is plotted. We will make a few different plots. Once you've selected your variables, click "Generate Plot" to render the figure. To make a new plot, change the parameters and click "Generate Plot" again. 

- Outcome Histogram
  - x-axis: Model: outcome
- Outocme Density Plot
  - x-axis: Model: outcome
  - plot type: density plot
- Scatterplot of Residuals vs Fitted Values
  - x-axis: Model: fitted.values
  - y-axis: Model: resid.marginal
  - plottype: scatterplot
  - Additional Options
    - Add y = 0 line
    - Add smooth line
- Boxplot of height by study
  - x-axis: Phenotype: study
  - y-axis: Phenotype: height
  - plot type: boxplot
- Boxplot of height by study, grouped by sex (sex indicated by color)
  - x-axis: Phenotype: study
  - y-axis: Phenotype: height
  - plot type: boxplot
  - group by: Phenotype: sex
- Boxplot of height by study, faceted by sex (each sex in its own panel)
  - x-axis: Phenotype: study
  - y-axis: Phenotype: height
  - plot type: boxplot
  - facet by: Phenotype: sex


#### Extracting Sample Genotypes from a GDS

Perhaps we want to look at the relationship between the genotype values of our association study "hits" and our phenotypes or model residuals. The GENESIS Model Explorer can do this as well if we provide the optional Genotype file with sample genotype values for the variants of interest. Conveniently, this file can be generated from an existing GDS file with the `GDS Genotype Extractor` application.

First, let's identify a few variants to use for this demonstration. After running an Application (e.g. GENESIS Single Variant Association Testing) on the BioData Catalyst Powered by Seven Bridges platform, the output files are saved in the directory `/sbgenomics/project-files/`. You can load these files into RStudio to explore them interactively -- we load the chromosome 8 single variant association test results from the task `07. 1KG Single Variant Association Test`, filter to variants with MAC $\geq 100$, and identify the 3 variants with the smallest p-values.

```{r, eval = FALSE}
library(data.table)

# load the association results and convert to a data.table
assoc <- data.table(get(load('/sbgenomics/project-files/1KG_height_single_chr8.RData')))

# filter to variants with MAC >= 100
assoc <- assoc[MAC >= 100]

# sort by p-value
setkey(assoc, Score.pval)
head(assoc)
```

The first 3 rows in the table are the variants with the smallest p-values. Note that they are not anywhere near genome-wide significant in our toy example, but we can still use them to illustrate the tools in this tutorial. 

We need to create a "variant include file" with the `variant.id` of these variants as input for the `GDS Genotype Extractor` application. The variant include should be saved as an .rds file using the `saveRDS` function. 

```{r, eval = FALSE}
varid <- assoc$variant.id[1:3]
varid

saveRDS(varid, file = '/sbgenomics/output-files/1KG_height_chr8_variant_include.rds')
```

We also need to create a "sample include file" with the `sample.id` of all the samples included in our analysis as input for the `GDS Genotype Extractor` application. We can get these `sample.id` values from our fitted null model -- we can use the `<output_prefix>_null_model_reportonly.RData` file, which is much smaller than the `<output_prefix>_null_model.RData` file by excluding some large matrices only needed for computing association test results. The sample include file should also be saved as an .rds file using the `saveRDS` function. 

```{r, eval = FALSE}
nullmod <- get(load('/sbgenomics/project-files/1KG_height_null_model_reportonly.RData'))

# the sample.id are stored in the "fit" data.frame
head(nullmod$fit)

sampid <- nullmod$fit$sample.id
length(sampid)

saveRDS(sampid, file = '/sbgenomics/output-files/1KG_height_sample_include.rds')
```

**Note about Directories:** The working directory for the Data Studio is `sbgenomics/workspace` (you can see this by going to the Terminal in RStudio and typing `pwd`). This directory is accessible in the Data Studio, but the Applications in your Project can **not** see files here. Applications only see the `sbgenomics/project-files/` directory (which is read-only from the Data Studio). In order to make our variant include file visible to the `GDS Genotype Extractor` application, we save our file to the `/sbgenomics/output-files/` directory. When we stop our RStudio session (and only then), new files in the `/sbgenomics/output-files/` directory will be copied over to the `/sbgenomics/project-files/` directory, making them available to Applications. More details can be found in the platform [documentation](https://sb-biodatacatalyst.readme.io/docs/about-files-in-a-data-cruncher-analysis).


### Exercise 5.1 (Application)

Use the `GDS Genotype Extractor` app on the BioData Catalyst powered by Seven Bridges platform to create an .rds file with genotype values for all samples in our `1KG_phase3_subset` GDS files at the 3 variants on chromosome 8 we identified above. The steps to perform this analysis are as follows:

- Copy the app to your project if it is not already there:
  - Click: Public Gallery > Apps > Browse
  - Search for `GDS Genotype Extractor`
  - Click: Copy > Select your project > Copy
- Run the analysis in your project:
  - Click: Apps > `GDS Genotype Extractor` > Run
  - Specify the Inputs:
    - GDS file: `1KG_phase3_subset_chr8.gds` 
    - Sample include file: `1KG_height_sample_include.rds` 
    - Variant include file: `1KG_height_chr8_variant_include.rds`
    - Output prefix: "1KG_height_chr8" (or any other string to name the output file)
  - Click: Run
  
The analysis will take a few minutes to run. You can find your analysis in the Tasks menu of your Project to check on its progress and see the results once it has completed.

The output of this analysis will be a `<output_prefix>_genotypes.rds` file that contains a column of `sample.id` and then one column per variant with the genotype values for each sample, and a `<output_prefix>_variant_info.rds` file with one row per variant and columns providing details such as variant identifiers, chromosome, position, and ref and alt alleles. 

You can find the expected output of this analysis by looking at the existing task `09. 1KG GDS Genotype Extractor chr8` in the Tasks menu of your Project. The output files are available in the Project, so you do not need to wait for your analysis to finish to move to the next exercise.


### Exercise 5.2 (GENESIS Model Explorer)

Use the GENESIS Model Explorer to make a boxplot of the Cholesky residuals (`resid.cholesky`) from the 1KG height null model by genotype value of the variant at chr8:73417176 G>A. The Cholesky residuals are a transformation of the marginal residuals computed using the estimated model covariance structure to remove the correlation among samples. The correlation of these residuals with the genotype values if essentially what the score statistic is measuring when we perform our association tests. What do you observe in the boxplot? 

### Solution 5.2 (GENESIS Model Explorer)

- In the GENESIS Model Explorer window, click back on the "Load Data" tab and add the following file:
  - Genotype File:
    - Project: select your SISG project
    - Current File: select `1KG_height_chr8_genotypes.rds` (this is the genotype file we created in Exercise 5.2)
  - Click: Load Data
- Set the plotting parameters as follows:
  - x-axis: Genotype: chr8:73417176_G_A
  - y-axis: Model: resid.cholesky
  - plot type: boxplot
  - Additional Options
      - Add y = 0 line
- Click "Generate plot"

From the boxplot, we can see that there is a slight "upward" trend in the median residual value across genotypes. The values 0/1/2 of the genotype value correspond to the number of copies of the alternate allele (in this case, the A allele), so we observe that having more copies of the A allele is associated with higher values of height, after adjusting for the covariates in our model. This is consistent with the `Score` value for this variant from our association test, which also has a positive value. 

NOTE: the trend across genotype values is not "steep" in this example. This is partially because our example variant is not very significant, but similar figures for genome-wide significant associations often won't look *that* different. Typically, for complex polygenic traits, the effect size of any one variant is quite small, so the overlap in residual distributions by genotype value is substantial, and the trend in the mean will be slight -- this is why GWAS need such large samples to have statistical power!


## LocusZoom Shiny App 

The [Locus Zoom Shiny App](https://locuszoom-shiny-app.bdc.sb-webapp.com/) is an interactive tool built on the [LocusZoom.js library](https://statgen.github.io/locuszoom/) that enables users to make LocusZoom plots of association results produced with the `GENESIS Single Variant Association Testing` app. We will now use the LocusZoom Shiny App to make a LocusZoom plot of our association "hit" on chr8.

- Launch the interactive browser
  - From the top menu, click "Public Resources" > "Interactive Web Apps"
  - Click: "Open" on the LocusZoom Shiny App
  - Click: "Yes" to proceed
  
The application requires data to be stored as a JSON file. There is a `GENESIS Data JSONizer` tool that converts single-variant association test results .RData file as output by the `GENESIS Single Variant Association Testing` app into the required JSON file. This tool also calculates the linkage disequilibrium (LD) measures required to make the LocusZoom plot for the selected variants.

- Click the "GENESIS Data JSONizer" tab at the top of the screen
- Select Input Files
  - GDS file: `1KG_phase3_subset_chr8.gds`
  - .RData file: `1KG_height_single_chr8.RData`
- JSONizer parameters
  - Check: "Specify variant and a flanking region around it"
  - Select the position of the variant of interest: 73417176
  - Specify flanking region: 1000000 (i.e. 1Mb in each direction. Note: our data is very sparse, you probably want a smaller flanking region with dense sequencing data)
  - Select test type: score
- Click: JSONize
  
You have the option to download the JSON file to your local environment or upload it to the BioData Catalyst platform and save it for later, if you desire. 

- Expand: JSON File - Download and Export Form
- Set a file name (e.g. "1KG_height_single_chr8_73417176")
- Choose extension: `.json`
- Click: Export JSON file to platform
- Select your Project and Click: Confirm
- Click: Upload

There are several optional data layers you can add to your LocusZoom plot. The most likely layer that you will want to adjust is the Linkage Disequilibrium (LD) layer. The tool gives you the option to either compute LD measures using your sample genotype data stored in the GDS file (the default), or use the University of Michigan (UM) database. 

- Expand: Option Data Layers
- Expand: Linkage Disequilibrium
- Select Data Source: Compute LD Data
- Select reference variant: 8:73417176_A/G (our variant of interest)
- Click: Calculate LD

You can expand the Linkage Disequilibrium Data Overview tab to see a preview of the calculated LD data, and you can download the data as a JSON file to your local environment or upload it to the BioData Catalyst platform and save it for later, if you desire.

- Expand: JSON File - Download and Export Form
- Set a file name (e.g. "1KG_height_single_chr8_73417176_LD")
- Choose extension: `.json`
- Click: Export JSON file to platform
- Select your Project and Click: Confirm
- Click: Upload

You need to select the Genome Build that matches your data:

- Change the Genome Build to GRCh37 for this dataset

You can review the Initial Plot State Info to make sure everything looks as expected, and then make the plot!

- Click: Generate plot

The generated plot is interactive. You can hover over variants to see their chromosome, position, alleles, and association p-value. You can drag the figure left or right to see different sections of the plotted region. You can save the current figure as a .png or .svg file either locally or on the BioData Catalyst platform. 

If you've saved your .json association results file and your .json LD statistics file to your Project, you can come back later and recreate your LocusZoom plot by selecting the "Use Your Own Data Sources" tab at the top of the LocusZoom Shiny App page. This time, rather than JSONizing the data, you can select the .json files as input, and set the plotting parameters the same as we did above.


## Exercise 5.3 (LocusZoom Shiny App)

Notably, the LocusZoom plot we generated with the example data is *very* sparse (there are only 7 variants), which is not representative of what a LocusZoom plot would actually look like in practice. There are example data sets available in the tool via the [University of Michigan database](https://portaldev.sph.umich.edu/docs/api/v1/#introduction). Select the "Explore UM Database" tab at the top of the LocusZoom Shiny App page and generate a LocusZoom plot using the GIANT Consortium BMI meta-analysis (PMID: 20935630) data for variant chr16:53803574, using a flanking region of 100kb. What is the p-value of the variant chr16:53803574_T/A? What gene is this variant located in? Change the LD reference population to EUR (European ancestry) -- what do you observe? Change the LD reference population to AFR (African ancestry) -- what do you observe? 

## Solution 5.3 (Locus Zoom Shiny App)

- The p-value of variant chr16:53803574_T/A is reported as $2.05 x 10^{-62}$
- This variant is located in the FTO gene, which is well established to be associated with BMI.
- Using the EUR LD reference panel, many of the variants in this region with similar p-values have very high LD with variant chr16:53803574_T/A (indicated by the red color).
- Using the AFR LD reference panel, many of the variants in this region with similar p-values no longer have high LD with variant chr16:53803574_T/A (indicated by the blue color).

